// ui/widget.cppï¼ˆå®Œæ•´ä¿®å¤ç‰ˆï¼‰
#include "widget.h"
#include "./ui_widget.h"
#include <QDebug>
#include "ClientCore.h" // åŒ…å«Coreå¤´æ–‡ä»¶
#include "ClientState.h" // åŒ…å«çŠ¶æ€è½¬æ¢å‡½æ•°

Widget::Widget(ClientCore* core, QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::Widget)
    , core_(core)  // åˆå§‹åŒ–CoreæŒ‡é’ˆ
{
    ui->setupUi(this);
    video_ = new VideoWidget(ui->videoContainer);
    adapter_ = new QtCameraAdapter(video_, this);

    // å…³é”®ä¿®å¤ï¼šç»‘å®šæ‘„åƒå¤´å›è°ƒåˆ°Qtç•Œé¢
    camera_.setFrameCallback([this](const Frame& f) {
        // è·¨çº¿ç¨‹å®‰å…¨è°ƒç”¨ï¼ˆCameraManageråœ¨å­çº¿ç¨‹ï¼ŒUIåœ¨ä¸»çº¿ç¨‹ï¼‰
        QMetaObject::invokeMethod(adapter_, "onFrame", Qt::QueuedConnection, Q_ARG(Frame, f));
    });

    if (core_) {
        core_->addListener(this);
    }

    // å…³é”®ä¿®å¤ï¼šç»‘å®šCoreäº‹ä»¶çš„è·¨çº¿ç¨‹å¤„ç†
    connect(this, &Widget::coreOutputReceived,
            this, &Widget::handleCoreOutput,
            Qt::QueuedConnection);
}

Widget::~Widget()
{
    // ä¿®å¤ï¼šä½¿ç”¨æ­£ç¡®çš„ç§»é™¤ç›‘å¬è€…æ–¹æ³•
    if (core_) {
        core_->removeListener(this);
    }
    delete ui;
}

// å®ç°ICoreListeneræ¥å£ï¼šæ¥æ”¶Coreè¾“å‡ºäº‹ä»¶
void Widget::onCoreOutput(const core::CoreOutput& out)
{
    qDebug() << "Main UI received core output, type index:" << out.index();
    // è½¬å‘ä¸ºQtä¿¡å·ï¼ˆç¡®ä¿UIçº¿ç¨‹å¤„ç†ï¼‰
    emit coreOutputReceived(out);
}

void Widget::handleCoreOutput(const core::CoreOutput& out)
{
    std::visit([this](auto&& e) {
        handle(e);   // ğŸ”¥ å’Œ LoginWidget ä¸€æ¨¡ä¸€æ ·çš„åˆ†å‘é£æ ¼
    }, out);
}

void Widget::handle(const core::OutOnlineUsers& e) {
    std::cout << "[UI] OutOnlineUsers" << std::endl;
    QString text;

    for (const auto& u : e.list) {
        text += QString("%1 (priv=%2)\n")
                    .arg(QString::fromStdString(u.name))
                    .arg(u.privilege);
    }

    ui->text_onlineUsers->setPlainText(text);
}
// // å®ç°æ§½å‡½æ•°ï¼šå¤„ç†Coreè¾“å‡ºäº‹ä»¶ï¼ˆUIçº¿ç¨‹ï¼‰
// void Widget::handleCoreOutput(const core::CoreOutput& out)
// {
//     // å¯æ ¹æ®éœ€è¦æ‰©å±•å¤„ç†é€»è¾‘ï¼Œç›®å‰æ‰“å°æ—¥å¿—
//     std::visit([](auto&& e) {
//         using T = std::decay_t<decltype(e)>;
//         if constexpr (std::is_same_v<T, core::OutStateChanged>) {
//             qDebug() << "Main UI state change:" << QString::fromStdString(stateToString(e.from))
//                      << "â†’" << QString::fromStdString(stateToString(e.to));
//         }
//     }, out);
// }

void Widget::on_Bt_video_on_off_clicked()
{
    qDebug("Button clicked!");
    camera_.start("/dev/video0");
}

void Widget::on_Bt_video_off_clicked()
{
    qDebug("Button clicked!");
    camera_.stop();
}

void Widget::on_Bt_tcp_send_clicked() {
    // è·å–ç›®æ ‡ç”¨æˆ·åå’Œæ¶ˆæ¯å†…å®¹
    QString target_user = ui->lineEdit_target_user->text();
    QString content = ui->lineEdit_msg->text();
    if (target_user.isEmpty() || content.isEmpty()) {
        std::cout << "can not be empty" << std::endl;
        return;
    }
    // å‘é€åˆ°Coreå±‚
    core_->postInput(core::InCmdSendText{
        target_user.toStdString(),
        content.toStdString()
    });
    // æ¸…ç©ºè¾“å…¥æ¡†
    ui->lineEdit_msg->clear();
}

// æ–°å¢ï¼šå¤„ç†å‘é€æ–‡æœ¬æ¶ˆæ¯ï¼ˆä»…æ—¥å¿—ï¼‰
void Widget::handle(const core::OutSendText& e) {
    std::cout << "[Widget] Send text to  " << e.target_user << ": " << e.content << std::endl;
}

// æ–°å¢ï¼šå¤„ç†æ¥æ”¶è½¬å‘æ–‡æœ¬æ¶ˆæ¯ï¼ˆæ˜¾ç¤ºåˆ°UIï¼‰
void Widget::handle(const core::OutForwardText& e) {
    std::cout << "[Widget] recev from " << e.from_user << " msg: " << e.content << std::endl;
    // è¿½åŠ åˆ°æ–‡æœ¬æ¡†
    QString text = ui->PTE_recv->toPlainText();
    text += QString("[%1]: %2\n").arg(QString::fromStdString(e.from_user)).arg(QString::fromStdString(e.content));
    ui->PTE_recv->setPlainText(text);
}

void Widget::handle(const core::OutStateChanged& e) {
    std::cout << "[Widget] FSM:"
             << (stateToString(e.from))
             << "â†’"
             << (stateToString(e.to)) << std::endl;


}

void Widget::handle(const core::OutDisconnected&) {
    std::cout << "[Widget] OutDisconnected" << std::endl;
}

void Widget::handle(const core::OutConnect&) {
    std::cout << "[Widget] Ignore OutConnect (CoreExecutor handles it)" << std::endl;
}

void Widget::handle(const core::OutSendLogin&) {
    std::cout << "[Widget] Ignore OutSendLogin" << std::endl;
}

void Widget::handle(const core::OutSendPing&) {
    std::cout << "[Widget] Ignore OutSendPing" << std::endl;
}

void Widget::handle(const core::OutUpdateAlive&) {
    std::cout << "[Widget] OutUpdateAlive" << std::endl;
}

void Widget::handle(const core::OutLoginOk&) {
    std::cout << "[Widget] OutLoginOk (optional handling)" << std::endl;
}

void Widget::handle(const core::OutLoginFail&) {
    std::cout << "[Widget] OutLoginFail (optional handling)" << std::endl;
}

void Widget::handle(const core::OutSelectLan&) {
    std::cout << "[Widget] OutSelectLan" << std::endl;
}

void Widget::handle(const core::OutSelectVpn&) {
    std::cout << "[Widget] OutSelectVpn)" << std::endl;
}
